scripts/keymigrate/migrate_test.go

```diff
@@ -0,0 +1,222 @@
+		"ConsensusParams":   []byte(fmt.Sprintf("consensusParamsKey:%d", val)),
+		"ABCIResponse":      []byte(fmt.Sprintf("abciResponsesKey:%d", val)),
+		"State":             []byte("stateKey"),
+		"CommittedEvidence": append([]byte{0x00}, []byte(fmt.Sprintf("%0.16X/%X", int64(val), []byte("committed")))...),
+		"PendingEvidence":   append([]byte{0x01}, []byte(fmt.Sprintf("%0.16X/%X", int64(val), []byte("pending")))...),
+		"LightBLock":        []byte(fmt.Sprintf("lb/foo/%020d", val)),
+		"Size":              []byte("size"),
+		"TxHeight":          []byte(fmt.Sprintf("tx.height/%s/%d/%d", fmt.Sprint(val), val, val)),
+		"UserKey0":          []byte(fmt.Sprintf("foo/bar/%d/%d", val, val)),
+		"UserKey1":          []byte(fmt.Sprintf("foo/bar/baz/%d/%d", val, val)),
+	}
+}
+
+func getNewPrefixKeys(t *testing.T, val int) map[string][]byte {
+	t.Helper()
+	return map[string][]byte{
+		"Height":            makeKey(t, int64(0), int64(val)),
+		"BlockPart":         makeKey(t, int64(1), int64(val), int64(val)),
+		"BlockPartTwo":      makeKey(t, int64(1), int64(val+2), int64(val+val)),
+		"BlockCommit":       makeKey(t, int64(2), int64(val)),
+		"SeenCommit":        makeKey(t, int64(3), int64(val)),
+		"BlockHeight":       makeKey(t, int64(4), int64(val)),
+		"Validators":        makeKey(t, int64(5), int64(val)),
+		"ConsensusParams":   makeKey(t, int64(6), int64(val)),
+		"ABCIResponse":      makeKey(t, int64(7), int64(val)),
+		"State":             makeKey(t, int64(8), int64(val)),
+		"CommittedEvidence": makeKey(t, int64(9), int64(val)),
+		"PendingEvidence":   makeKey(t, int64(10), int64(val)),
+		"LightBLock":        makeKey(t, int64(11), int64(val)),
+		"Size":              makeKey(t, int64(12)),
+		"UserKey0":          makeKey(t, "foo", "bar", int64(val), int64(val)),
+		"UserKey1":          makeKey(t, "foo", "bar/baz", int64(val), int64(val)),
+		"TxHeight":          makeKey(t, "tx.height", fmt.Sprint(val), int64(val), int64(val+2), int64(val+val)),
+	}
+}
+
+func getLegacyDatabase(t *testing.T) (int, dbm.DB) {
+	db := dbm.NewMemDB()
+	batch := db.NewBatch()
+	ct := 0
+
+	generated := []map[string][]byte{
+		getLegacyPrefixKeys(2),
+		getLegacyPrefixKeys(9001),
+		getLegacyPrefixKeys(math.MaxInt32),
+		getLegacyPrefixKeys(math.MaxInt64 - 8),
+	}
+
+	// populate database
+	for _, km := range generated {
+		for _, key := range km {
+			ct++
+			require.NoError(t, batch.Set(key, []byte(fmt.Sprintf(`{"value": %d}`, ct))))
+		}
+	}
+	require.NoError(t, batch.WriteSync())
+	require.NoError(t, batch.Close())
+	return ct - (2 * len(generated)) + 2, db
+}
+
+func TestMigration(t *testing.T) {
+	t.Run("Idempotency", func(t *testing.T) {
+		// we want to make sure that the key space for new and
+		// legacy keys are entirely non-overlapping.
+
+		legacyPrefixes := getLegacyPrefixKeys(42)
+
+		newPrefixes := getNewPrefixKeys(t, 42)
+
+		require.Equal(t, len(legacyPrefixes), len(newPrefixes))
+
+		t.Run("LegacyEvidence", func(t *testing.T) {
```

Comment:
Why do you have the Evidence suffix here? This is testing all keys right? not just evidence ones